File: Sources/IO/Handlers/HandleTab.swift
func handleTab(_ sequence: String, _ prompt: inout PromptState, _ session: borrowing SessionState) -> InputAction {
    let completed = tabComplete(fuzz: prompt.content, state: session)
    if completed != prompt.content {
        prompt.content += completed
        print(completed, terminator: "")
    }

	return InputAction.continueReading
}

func tabComplete(fuzz: String, state: borrowing SessionState) -> String {
    if fuzz.count == 0 {
        return ""
    }
    if fuzz.split(separator: " ").count == 1 && fuzz.last != " " {
        let startsWith = state.availableCommands.filter{$0.starts(with: fuzz)}
        if(startsWith.count > 0) {
            return startsWith.first!.replacingOccurrences(of: fuzz, with: "")
        }
    } else if fuzz.split(separator: " ").count > 1 || fuzz.last == " " {
        //to be continued, this needs to autocomplete paths
    }
    
    return fuzz
}

File: src/io/handlers/handle_tab.rs
use crate::models::prompt_state::PromptState;
use crate::core::session_state::SessionState;
use crate::io::handler_registry::InputAction;

pub fn handle_tab(_sequence: &str, prompt: &mut PromptState, session: &SessionState) -> InputAction {
    let completed = tab_complete(&prompt.content, session);
    if completed != prompt.content {
        prompt.content.push_str(&completed);
        print!("{}", completed);
    }
    InputAction::ContinueReading
}

fn tab_complete(fuzz: &str, state: &SessionState) -> String {
    if fuzz.is_empty() {
        return String::new();
    }

    let parts: Vec<&str> = fuzz.split_whitespace().collect();
    if parts.len() == 1 && !fuzz.ends_with(' ') {
        let starts_with: Vec<&String> = state
            .available_commands
            .iter()
            .filter(|cmd| cmd.starts_with(fuzz))
            .collect();
        if !starts_with.is_empty() {
            return starts_with[0].replacen(fuzz, "", 1);
        }
    }
    // Path completion to be continued

    fuzz.to_string()
}
